src/group_theory/group_action/defs.lean

```diff
@@ -778,3 +759,45 @@ See note [reducible non-instances]. -/
 @[reducible]
 def add_action.of_End_hom [add_monoid M] (f : M →+ additive (function.End α)) : add_action M α :=
 add_action.comp_hom α f
+
+/-! ### `additive`, `multiplicative` -/
+
+section
+
+open additive multiplicative
+
+instance additive.has_vadd [has_smul α β] : has_vadd (additive α) β := ⟨λ a, (•) (to_mul a)⟩
+instance multiplicative.has_smul [has_vadd α β] : has_smul (multiplicative α) β :=
+⟨λ a, (+ᵥ) (to_mul a)⟩
+
+@[simp] lemma to_mul_smul [has_smul α β] (a) (b : β) : (to_mul a : α) • b = a +ᵥ b := rfl
+@[simp] lemma of_mul_vadd [has_smul α β] (a : α) (b : β) : of_mul a +ᵥ b = a • b := rfl
+@[simp] lemma to_add_vadd [has_vadd α β] (a) (b : β) : (to_add a : α) +ᵥ b = a • b := rfl
+@[simp] lemma of_add_smul [has_vadd α β] (a : α) (b : β) : of_add a • b = a +ᵥ b := rfl
```

Comment:
I'm not convinced that the `to` versions (which are the ones we didn't have before) are helpful as simp lemmas.

Reply:
Which is to say; I would consider stating them backwards and writing them as `additive.vadd_def` or similar to match `sum.smul_def` or similar; but I could be persuaded otherwise.

Reply:
I have had success using similar lemmas in the past. My reasoning is that there's no preferred type from the point of view of `simp`: `additive α` is a type synonym of `α`, but `α` is also a type synonym of `additive α`. And indeed this seems to keep the simp set confluent as for example `to_mul (of_mul a) • b` can be rewritten either directly as `a • b` or as `of_mul a • b` then `a • b`.

It's completely analogous (save for arity) to how we have both `of_add_add : of_add (a + b) = of_add a * of_add b` and `to_add_mul : to_add (a * b) = to_add a + to_add b` but do not have `multiplicative.mul_def : a * b = of_add (to_add a + to_add b)`.

I will nevertheless add that it seems we have the `_def` versions for *some* unary operations, so I will add them if you insist.

Reply:
> It's completely analogous (save for arity) to how we have both `of_add_add : of_add (a + b) = of_add a * of_add b` and `to_add_mul : to_add (a * b) = to_add a + to_add b` but do not have `multiplicative.mul_def : a * b = of_add (to_add a + to_add b)`.

`to_add_mul` has the `to_add` on the outside of the expression on the LHS, and on the inside on the RHS. By this analogy, the direction should be `vadd_def : a +ᵥ b = to_mul a • b` (which is useful because it is likely that a lemma about the action of `α` on `β` can then be applied).

> And indeed this seems to keep the simp set confluent as for example `to_mul (of_mul a) • b` can be rewritten either directly as `a • b` or as `of_mul a +ᵥ b` then `a • b`.

I don't see your point here; doesn't my suggested `vadd_def` lemma also give a confluent simp set, whether it's `simp` or not?

Reply:
> `to_add_mul` has the `to_add` on the outside of the expression on the LHS, and on the inside on the RHS. By this analogy, the direction should be `vadd_def : a +ᵥ b = to_mul a • b`.

... except that there is no `to_mul` on the LHS. `vadd_self` is effectively introducing a function application that didn't exist before.